Quick_Deploy/vLLM/model_repository/vllm/1/model.py

```diff
@@ -0,0 +1,330 @@
+
+    completions: List[Completion] = []
+
+    # inflight request stats
+    current_inflight_count: Optional[int] = None
+    average_inflight_count: Optional[float] = None
+
+
+class OutputOptions:
+    def __init__(self, *, echo: bool = False, stream: bool = False):
+        self._echo = echo
+        self._stream = stream
+
+    @property
+    def echo(self) -> bool:
+        """
+        If true, the prompt and prompt token ids will be echoed back to the client in the response.
+        """
+        return self._echo
+
+    @property
+    def stream(self) -> bool:
+        """
+        If true, the response will be streamed back to the client as it is generated.
+        """
+        return self._stream
+
+
+class TritonPythonModel:
+    def initialize(self, args):
+        self.logger = pb_utils.Logger
+
+        # load triton model config from args
+        self.model_config = json.loads(args["model_config"])
+
+        # assert are in decoupled mode
+        self.using_decoupled = pb_utils.using_decoupled_model_transaction_policy(
+            self.model_config
+        )
+        assert (
+            self.using_decoupled
+        ), "vLLM Triton backend must be configured to use decoupled model transaction policy"
+
+        engine_args_filepath = os.path.join(
+            args["model_repository"], _VLLM_ENGINE_ARGS_FILENAME
+        )
+        assert os.path.isfile(
+            engine_args_filepath
+        ), f"'{_VLLM_ENGINE_ARGS_FILENAME}' containing vllm engine args must be provided in '{args['model_repository']}'"
+        with open(engine_args_filepath) as file:
+            vllm_engine_config = VLLMAsyncEngineConfig(**json.load(file))
+
+        # create the vllm engine
+        engine_args = AsyncEngineArgs(
+            vllm_engine_config.model,
+            disable_log_requests=vllm_engine_config.disable_log_requests,
+            max_num_batched_tokens=vllm_engine_config.max_num_batched_tokens,
+            max_num_seqs=vllm_engine_config.max_num_seqs,
+        )
+        self.llm = AsyncLLMEngine.from_engine_args(engine_args)
+
+        self._loop = asyncio.get_event_loop()
+        self._loop_thread = threading.Thread(
+            target=self.engine_loop, args=(self._loop,)
+        )
+        self._shutdown_event = asyncio.Event()
+        self._loop_thread.start()
+
+        # local counters to track inflight requests
+        self._inflight_counter = 0
+
+    def create_task(self, coro):
+        """
+        Creates a task on the engine's event loop which is running on a separate thread.
+        """
+        assert (
+            self._shutdown_event.is_set() is False
+        ), "Cannot create tasks after shutdown has been requested"
+
+        async def _wrapped_coro(coro):
+            """
+            Wraps the given coroutine in a new coroutine and decrements the in-flight counter after awaiting the original coroutine.
+            """
+            await coro
+            self._inflight_counter -= 1
```

Comment:
instead of the wrapped co-routine - can we add the inflight counter -- to the generate method itself? so generate would increment and decrement directly?